### PR TITLE
[PhpUnitBridge] Make ClockMock match namespaces that begin with Tests\\

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q | A |
| --- | --- |
| Branch? | 2.8 |
| Bug fix? | yes? |
| New feature? | no |
| BC breaks? | no? |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | symfony/symfony-docs#6546 |

See symfony/symfony-standard#969

### CODE REVIEW

File: src/Symfony/Bridge/PhpUnit/ClockMock.php [hunk 1]
```diff
@@ -74,9 +74,10 @@ public static function register($class)
         $self = get_called_class();
 
         $mockedNs = array(substr($class, 0, strrpos($class, '\\')));
-        if (strpos($class, '\\Tests\\')) {
-            $ns = str_replace('\\Tests\\', '\\', $class);
-            $mockedNs[] = substr($ns, 0, strrpos($ns, '\\'));
+        $count = 0;
+        $ns = preg_replace('/(^|\\\\)Tests\\\\/', '$1', $mockedNs[0], 1, $count);
+        if (1 === $count) {
+            $mockedNs[] = $ns;
```
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: imho we should simply additionally check if the namespace starts with `Tests\`:

```php
if (false !== strpos($class, '\\Tests\\')) {
    // ...
} elseif (0 === strpos('Tests\\')) {
    $mockedNs[] = substr($class, 6);
}
```

Comment (teohhanhui) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: Using regex seems more extensible to me, even though it might not be as readable...

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: I'd favor readability here, thus @xabbuh proposal

**Comment by teohhanhui** [comment]
@nicolas-grekas: Addressed @xabbuh's comment.

File: src/Symfony/Bridge/PhpUnit/ClockMock.php [hunk 2]
```diff
@@ -73,10 +73,12 @@ public static function register($class)
...
         $self = get_called_class();
 
-        $mockedNs = array(substr($class, 0, strrpos($class, '\\')));
-        if (strpos($class, '\\Tests\\')) {
-            $ns = str_replace('\\Tests\\', '\\', $class);
-            $mockedNs[] = substr($ns, 0, strrpos($ns, '\\'));
+        $testNs = substr($class, 0, strrpos($class, '\\'));
+        $mockedNs = array($testNs);
+        if (false !== ($pos = strpos($testNs, '\\Tests\\'))) {
+            $mockedNs[] = substr_replace($testNs, '\\', $pos, 7);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: this changes the semantic a bit (only the first `\Tests\` is replaced now). Let's be safe and keep the code asis for this first part...

Comment (teohhanhui) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: Wouldn't that actually have been a bug? Or was that the intended behaviour?

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: not a bug (nor really planned but we shouldn't change now)

Comment (teohhanhui) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: Done

**Comment by teohhanhui** [comment]
I believe Travis failure is unrelated.

File: src/Symfony/Bridge/PhpUnit/ClockMock.php [hunk 3]
```diff
@@ -74,9 +74,11 @@ public static function register($class)
         $self = get_called_class();
 
         $mockedNs = array(substr($class, 0, strrpos($class, '\\')));
-        if (strpos($class, '\\Tests\\')) {
+        if (false !== strpos($class, '\\Tests\\')) {
             $ns = str_replace('\\Tests\\', '\\', $class);
             $mockedNs[] = substr($ns, 0, strrpos($ns, '\\'));
+        } elseif (0 === strpos($class, 'Tests\\')) {
+            $mockedNs[] = substr($class, 6, strrpos($class, '\\'));
```
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ClockMock.php]: Must be:

```php
$mockedNs[] = substr($class, 6, strrpos($class, '\\') - 6);
```

**Comment by teohhanhui** [comment]
Status: Needs Review

**Comment by xabbuh** [comment]
:+1:

Status: Reviewed

**Comment by nicolas-grekas** [comment]
feature on 3.2 isn't it?

**Comment by teohhanhui** [comment]
> feature on 3.2 isn't it?

If so, then we need to address this issue in symfony-standard, because it does not work as expected out of the box.

**Comment by nicolas-grekas** [comment]
there is no deps bewteen the bridge and sfstd, so no need to change, just use the newest bridge to run the tests

**Comment by teohhanhui** [comment]
> just use the newest bridge to run the tests

That means bumping the version of the PhpUnit Bridge in symfony-standard 2.8?

**Comment by nicolas-grekas** [comment]
Thank you @teohhanhui.

**Pull request closed without merging**